std/string.d

```diff
@@ -3171,108 +3171,269 @@ S detab(S)(S s, size_t tabSize = 8) @trusted pure
     Params:
         s       = String to convert.
         tabSize = Tab columns are $(D tabSize) spaces apart.
+
+    Returns:
+        GC allocated string with tabs replaced with spaces;
+        use $(LREF entabber) to not allocate.
+
+    See_Also:
+        $(LREF entabber)
  +/
-S entab(S)(S s, size_t tabSize = 8) @trusted pure
+S entab(S)(S s, size_t tabSize = 8) @trusted
     if (isSomeString!S)
 {
-    import std.utf : encode;
-    import std.uni : lineSep, paraSep;
-    import std.exception : assumeUnique;
+    import std.array;
+    return cast(S)(entabber(s, tabSize).array);
+}
 
-    bool changes = false;
-    alias C = Unqual!(typeof(s[0]));
-    C[] result;
+///
+unittest
+{
+    assert(entab("        x \n") == "\tx\n");
+}
+
+/++
+    Replaces spaces in range $(D r) with the optimal number of tabs.
+    All spaces and tabs at the end of a line are removed.
 
-    int nspaces = 0;
-    int nwhite = 0;
-    size_t column = 0;         // column number
+    Params:
+        r = string or forward range
+        tabSize = distance between tab stops
 
-    foreach (size_t i, dchar c; s)
+    Returns:
+        lazy forward range with spaces replaced with tabs
+
+    See_Also:
+        $(LREF entab)
+  +/
+auto entabber(Range)(Range r, size_t tabSize = 8)
+    if (isForwardRange!Range)
+{
+    import std.uni : lineSep, paraSep, nelSep;
+    import std.utf : codeUnitLimit, decodeFront;
+
+    assert(tabSize > 0);
+    alias C = Unqual!(ElementEncodingType!Range);
+
+    static struct Result
     {
+    private:
+        Range _input;
+        size_t _tabSize;
+        size_t nspaces;
+        size_t ntabs;
+        int column;
+        size_t index;
 
-        void change()
+        @property C getFront()
         {
-            changes = true;
-            result = null;
-            result.length = s.length;
-            result.length = i;
-            result[0 .. i] = s[0 .. i];
+            static if (isSomeString!Range)
+                return _input[0];       // avoid autodecode
```

Comment:
That's problematic because, because you're disabling decoding for the user of entabber, but still present a dchar range.